timeseries/src/autogluon/timeseries/dataset/ts_dataframe.py

```diff
@@ -634,3 +634,131 @@ def get_reindexed_view(self, freq: str = "S") -> TimeSeriesDataFrame:
         df_view._cached_freq = freq
 
         return df_view
+
+    def to_regular_index(self, freq: str) -> "TimeSeriesDataFrame":
+        """Fill the gaps in an irregularly-sampled time series with NaNs.
+
+        Parameters
+        ----------
+        freq: str
+            Frequency string of the new time series data index.
+
+        Examples
+        --------
+        >>> print(ts_dataframe)
+                            target
+        item_id timestamp
+        0       2019-01-01     NaN
+                2019-01-03     1.0
+                2019-01-06     2.0
+                2019-01-07     NaN
+        1       2019-02-04     3.0
+                2019-02-07     4.0
+
+        >>> print(ts_dataframe.to_regular_index(freq="D"))
+                            target
+        item_id timestamp
+        0       2019-01-01     NaN
+                2019-01-02     NaN
+                2019-01-03     1.0
+                2019-01-04     NaN
+                2019-01-05     NaN
+                2019-01-06     2.0
+                2019-01-07     NaN
+        1       2019-02-04     3.0
+                2019-02-05     NaN
+                2019-02-06     NaN
+                2019-02-07     4.0
+
+        """
+        if self.freq is not None:
+            if self.freq != freq:
+                raise ValueError(
+                    f"TimeSeriesDataFrame already has a regular index with freq '{self.freq}' "
+                    f"that cannot be converted to the given freq '{freq}'"
+                )
+            else:
+                return self
+
+        filled_series = []
+        for item_id, time_series in self.groupby(level=ITEMID, sort=False):
+            time_series = time_series.droplevel(ITEMID)
+            timestamps = time_series.index
+            if not timestamps.is_monotonic_increasing:
+                raise ValueError(
+                    "Please make sure that the timestamps are sorted in increasing order for all time series."
+                )
+            start = timestamps[0]
+            end = timestamps[-1]
+            filled_timestamps = pd.date_range(start=start, end=end, freq=freq, name=TIMESTAMP)
```

Comment:
For simplicity, can we use `resample` [function](https://pandas.pydata.org/docs/reference/api/pandas.Series.resample.html)?

Reply:
This is really cool and feels a lot more elegant, thanks!